File: Scala/src/main/scala/AlgorithmicHeights/BINS.scala
package AlgorithmicHeights

object Bins {
  import scala.annotation.tailrec

  private def convertToIntVector(line: String): Vector[Int] = line.split(" ").map(_.toInt).toVector

  def findElemsInSortedArray(array: Vector[Int], queries: Vector[Int], n: Int): Vector[Int] = {
    @tailrec
    def binarySearch(item: Int, left: Int, right: Int): Int = {
      if (left > right) -1
      else {
        val middleIx: Int = (left + right) / 2
        val middleElem: Int = array(middleIx)
        if (middleElem == item) middleIx + 1
        else if (middleElem < item) binarySearch(item, middleIx + 1, right)
        else binarySearch(item, left, middleIx - 1)
      }
    }

    queries.map(binarySearch(_, 0, n - 1))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val _: Int = reader.next().toInt
    val array: Vector[Int] = convertToIntVector(reader.next())
    val queries: Vector[Int] = convertToIntVector(reader.next())
    val result: Vector[Int] = findElemsInSortedArray(array, queries, n)
    println(result.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/bins.ml
let convert_to_ints line =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let find_elems_in_sorted_array (array : int array) (queries : int array) (n : int) : int array =
    let rec binary_search item left right =
        if left > right then -1
        else
            let middle_ix = (left + right) / 2 in
            let middle_elem = array.(middle_ix) in
            if middle_elem = item then middle_ix + 1
            else if middle_elem < item then binary_search item (middle_ix + 1) right
            else binary_search item left (middle_ix - 1) in
    Array.map (fun x -> binary_search x 0 (n - 1)) queries


let () =
    let n = read_int () in
    let _ = read_int () in
    let array = convert_to_ints (read_line ()) in
    let queries = convert_to_ints (read_line ()) in
    let results = find_elems_in_sorted_array array queries n in
    Array.(results |> map string_of_int |> to_list) |> String.concat " " |> print_endline
